Use GtkBuilder to create App Menu

nickleus27 opened the pull request:
This is a draft to update deprecated ```Gtk::UIManager``` methods to create the applications menu bar.  Instead this pull request uses ```Gtk::Builder``` to creae application menu bar.  This is currently a work in progress.


UPD:
Discussion start here: https://github.com/synfig/synfig/pull/2498#issuecomment-1054922175

nickleus27 commented:
@ice0 @rodolforg This is a draft for the new menu bar.

ice0 commented:
@nickleus27 
You can switch PR to draft to show it's WIP.

![Screenshot_22](https://user-images.githubusercontent.com/5604544/159652228-59fd9bc6-e596-4ccc-b037-dfda423598b5.png)

nickleus27 commented:
@ice0 @rodolforg  Starting at ```L873``` in changed files I added some comments ```//TODO``` . First do you have an idea of how to add menu items that is not permanent like ```Open Recent``` or point me in a direction to learn about this.  Second I am having a hard time finding some action call backs that I put in ```//TODO```. Third, i am having trouble finding path to ```.png``` for custom icons. Thank you for your time and help on this issue.

ice0 commented:
> //TODO1: ADD Open Recent Actions

There are two different approaches:
1) Because `ui_info` is just a string, you can add new elements in place. Something like this (pseudocode): 
```
Glib::ustring ui_info = "...";

// Add recent items
for (const auto recent_item : recent_list) {
  ui_info += 	"        <item>" +
    etl::strprintf("          <attribute name='label' translatable='no'>%s</attribute>", recent_item.label) +
    "        </item>";
}
// following menu items
ui_info += "...";
```

This is not the best idea because you need to rebuild the main menu every time the dynamic element changes. 

2) Attach new elements at runtime. This is more preferable.
This is currently done in the `MainWindow::on_recent_files_changed()` method. It uses the ability of the UIManager class to merge UI elements.

And it looks like the same approach can be used with `Gtk::Builder`: 
> bool Gtk::Builder::add_from_string(const Glib::ustring & buffer)
> Parses a string containing a GtkBuilder UI definition and merges it with the current contents of the builder.

> //TODO2: need to find cb for save-all

https://github.com/synfig/synfig/blob/042c7378917c32af4354bac5d35283a63a7b828c/synfig-studio/src/gui/canvasview.cpp#L1475-L1477

It calls `void CanvasView::save_all()`.

> //TODO3: need to find cb for render

Same here:

https://github.com/synfig/synfig/blob/042c7378917c32af4354bac5d35283a63a7b828c/synfig-studio/src/gui/canvasview.cpp#L1487-L1489

nickleus27 commented:
@ice0 I see how recent files can be added dynamically in the method ``` MainWindow::on_recent_files_changed()``` I could add code there to add recent files into the ```Gtk::Builder``` My current problem is I am unsure how to have my elements to merge correctly into the xml string.  I tried a test example at L997 in ```App::init_menu_builder()``` However, in the try catch block where I call ```add_from_string``` it throws an error. The string does not merge correctly. Could you look at my xml string and see what I need to do to get it to add items into the submenu with ```id=`recent```` . the submenu is at L911

Second, I when I click on ```Close Document``` twice without opening a new document the app crashes and I get this error:
```Assertion failed: (obj), function operator->, file _handle.h, line 740.
zsh: abort      ./output/Debug/bin/synfigstudio```

I also noticed that the ```Close-Ducment``` label does not get greyed out (cannot click) after closing the instance like it does on the old menubar.

rodolforg commented:
When using the new GMenu, dynamic menus should be implemented via Gio::MenuModel / Gio::Menu.
So the Open Recent menu specifically should use it explicitly.
Take a look at https://developer-old.gnome.org/glibmm/stable/classGio_1_1MenuModel.html#a0b0c42bcd34312965430750a2f56071a and https://developer-old.gnome.org/glibmm/stable/classGio_1_1MenuModel.html#afd1b2461f5e75317ce6fa58a39351870 `(Gio::MenuModel::items_changed()` and `Gio::MenuModel::signal_items_changed())`

rodolforg commented:
> I also noticed that the `Close-Ducment` label does not get greyed out (cannot click) after closing the instance like it does on the old menubar.

They are (currently) controlled by an ActionGroup of the active CanvasView. Check `CanvasView::activate()`, `CanvasView::deactivate()`, `CanvasView::~CanvasView()`.

nickleus27 commented:
+1

nickleus27 commented:
I have files adding dynamically now to Open Recent submenu. But duplicate files will be added to the list.  It seems
```	// remove group if exists
	typedef std::vector< Glib::RefPtr<Gtk::ActionGroup> > ActionGroupList;
	ActionGroupList groups = App::ui_manager()->get_action_groups();
	for(ActionGroupList::const_iterator i = groups.begin(); i != groups.end(); ++i)
		if ((*i)->get_name() == action_group->get_name())
			App::ui_manager()->remove_action_group(*i);
	groups.clear();
```
might take care of this in ```MainWindow::on_recent_file_changed()```? If so, what would be an equivalent to ```get_action_groups```? Thank you.

ice0 commented:
> My current problem is I am unsure how to have my elements to merge correctly into the xml string.

I also did not find a solution using the `GtkBuilder`, so I'll ask the GTK team about that. 

> Here is where I think duplicate files are prevented from going into menu:

It looks like removing duplicates in action groups not in menus.

> I see `Gio::Menu` has a `remove` method, but I am not sure how to iterate menu list for duplicates. 

Since you are adding all recent items every time there is no need to remove duplicates, you can just remove all items in the menu using `remove_all()` method. Something like this:

```
auto menu_object = App::menu_builder()->get_object("recent-file");
auto recent_file_menu = Glib::RefPtr<Gio::Menu>::cast_dynamic(menu_object);
recent_file_menu.remove_all();
```

nickleus27 commented:
@ice0 
>It looks like removing duplicates in action groups not in menus.

Do I need to worry about removing duplicate actions? I looked into this and I could not find a way to insert ```Gio::SimpleActionGroup``` into ```Gio::Application``` in a similar fashion that is already done with:
>void
MainWindow::on_recent_files_changed()
{
	// TODO(ice0): switch to GtkRecentChooserMenu?
	Glib::RefPtr<Gtk::ActionGroup> action_group = Gtk::ActionGroup::create("mainwindow-recentfiles");

 But it seems that ```Gio::Application```and ```Gio::SimpleActionGroup``` inherit from ```Gio::ActionMap``` interface which I believe is a hash map so adding the action again would just map to the same location?

nickleus27 commented:
are the icon images in ```synfig-studio/images/```? I have not been able to get them to work with the ```"<attribute name='icon'>"``` in the xml string.

nickleus27 commented:
@ice0 
>Use Gio::ActionGroup and Gtk::Widget::insert_action_group() instead.

Thank you. However, ``` Gtk::Widget::insert_action_group()``` and ``` Gtk::Widget::get_action_group``` is compiling, but causing me segfaults when I run the program.

nickleus27 commented:
@ice0 thank you. Looks like I might need to go with glade instead of .xml

nickleus27 commented:
I cannot seem to get .glade to work, and I cannot seem to get the widget I need for inserting action groups with the .xml I have that does work.

ice0 commented:
> If i am using `glade` then how do i `set_menubar` which takes a `gio::menu` as a parameter. Glade autogenerates the class objects different from `Gio::Menu`.

I just want to say that I need time to test this. I'll be back when there's news.

nickleus27 commented:
@ice0 @rodolforg
Thank you ice for looking into ```glade``` for me. 

 I am at a wall at this point.  I have not been able to find a way to get a widget that allows me to ```insert_action_group``` which seems to be needed for many reasons.  I need to attach and detach ```action_groups``` in many places it seems. For example in ```mainwindow.cpp``` and ```layeractionmanager.cpp```. Without a way to get a widget that is connected to the ```menu_bar``` I cannot remove a group of actions  or insert a group. I do not know how to continue.  Otherwise the only work around I have found is to keep a reference to each individual ```Gio::SimpleAction```. Then I could use ```set_enable(bool)``` which is similar to ```set_sensitive```.

nickleus27 commented:
I found that you can 
```static Gtk::MenuBar* menu_bar = Gtk::manage(new Gtk::MenuBar(gmenu));```
then pack it in a ```Gtk::Box``` in ```MainWindow constructor```. This keeps access to the ```menubar widget``` for inserting groups of actions, but the menubar does not display in the native ```macos menubar```.

nickleus27 commented:
@ice0 @rodolforg I have come up with a work arround for adding and removing action groups.
```typedef std::vector< Glib::RefPtr<Gio::SimpleAction> > SimpActionGroup;```
and
```
App::enable_action_group(bool isEnabled){
	for(auto action : App::canvas_action_group())
		action->set_enabled(isEnabled);
}
```
This way I can push actions into ```SimpActionGroup``` vector and use set_enabled to turn the actions on and off. Do you think this is a good way of doing it?

ice0 commented on synfig-studio/src/gui/app.cpp:
```diff
@@ -861,6 +865,274 @@ class Preferences : public synfigapp::Settings
+	//icon path
+	//this will only work if executing program from cmake-build
+	//this needs to be fixed
+	std::string icon_path = std::__fs::filesystem::current_path();
```
Use `studio::ResourceHelper::get_icon_path()` for this.

ice0 commented on synfig-studio/src/gui/app.h:
```diff
@@ -84,6 +85,9 @@ class Preferences;
 
 typedef Gtk::UIManager UIManager;
+typedef Gtk::Builder Builder;
+typedef std::list< Glib::RefPtr<Gio::SimpleAction> > MenuActionGroup;
```
Sorry if I miss something. 
Why not use `Gio::SimpleActionGroup` instead?

nickleus27 replied:
@ice0 thank you for noticing. I was unsure how to iterate through the actions with ```Gio::SimpleActionGroup```. How do you use an iterator with ```Gio::SimpleActioGroup``` to iterate through the actions for ```set_enable```?

ice0 replied:
```
auto group = Gio::SimpleActionGroup::create();
for (const auto& action_name : group->list_actions()) {
	auto action = group->lookup_action(action_name);
	// convert Gio::Action to Gio::SimpleAction to have access to `set_enabled()` method
	if (auto simpleAction = Glib::RefPtr<Gio::SimpleAction>::cast_dynamic(action)) {
		simpleAction->set_enabled(false);
	}		
}
```

ice0 commented:
Here is what I found: 

Good description about `Gio::MenuModel` is here: https://developer-old.gnome.org/glibmm/stable/classGio_1_1MenuModel.html#details

About `app` and `win` prefixes.
> E.g. when the model is exported as the application menu of a Gtk::Application, actions can be application-wide or window-specific (and thus come from two different action groups). By convention, the application-wide actions have names that start with "app.", while the names of window-specific actions start with "win.".


> Without a way to get a widget that is connected to the `menu_bar` I cannot remove a group of actions or insert a group. I do not know how to continue. Otherwise the only work around I have found is to keep a reference to each individual `Gio::SimpleAction`. Then I could use `set_enable(bool)` which is similar to `set_sensitive`.

Yes. This is the same way as it currently done. 

In a place where you don't want to save references, you can do something like this: 
```
	auto action = app->add_action("quit");
	...
	auto quit_action = app->lookup_action("quit");
	// convert Gio::Action to Gio::SimpleAction to have access to `set_enabled()` method
	if (auto action = Glib::RefPtr<Gio::SimpleAction>::cast_dynamic(quit_action)) {
		action->set_enabled(false);
	}
```

I'm not sure I've answered all the questions, so feel free to point them out.

nickleus27 commented:
@ice0 could you review ```layeractionmanager.cpp```. I know the ```cut``` action is not working correctly. Thank you for your time.

ice0 commented:
> @ice0 could you review `layeractionmanager.cpp`. I know the `cut` action is not working correctly. Thank you for your time.

Just checked and it works fine. Am I missing something?

nickleus27 commented:
> > @ice0 could you review `layeractionmanager.cpp`. I know the `cut` action is not working correctly. Thank you for your time.
> 
> Just checked and it works fine. Am I missing something?

When I use ```cut``` from the menubar I am creating it works differently then from old menubar.  The first cut i make doesn't cut the layer out like the old menubar does. Did you not experience that?

ice0 commented:
> When I use `cut` from the menubar I am creating it works differently then from old menubar. The first cut i make doesn't cut the layer out like the old menubar does. Did you not experience that? The following cuts seem to work right afterwards. It works intermittently.

Very strange. Can you make a video showing the process?
What I've done:
1) Launched Synfig Studio
2) Added a `Circle` layer
3) Edit->Cut

Everything works fine.

nickleus27 commented:
> > When I use `cut` from the menubar I am creating it works differently then from old menubar. The first cut i make doesn't cut the layer out like the old menubar does. Did you not experience that? The following cuts seem to work right afterwards. It works intermittently.
> 
> Very strange. Can you make a video showing the process? What I've done:
> 
> 1. Launched Synfig Studio
> 2. Added a `Circle` layer
> 3. Edit->Cut
> 
> Everything works fine.

After making recording I realize you are right. Thank you for checking.

nickleus27 commented:
@ice0 When you have a chance could you check to see if the toggle actions in the `Show/Hide Handles` submenu are working correctly.  I am unsure if I have them toggling correctly.  Also, if you could look at the `.xml` string called `ui_info` in the `init_menu_builder` function in `app.cpp`. I have `TODO` comments by accelerators that are not parsing correctly.  Not very familiar with xml notation. Thank you.

nickleus27 commented:
@ice0 @rodolforg  Hey guys, whats up? I am currently working on the `Toolbox` submenu in the menubar. So it looks like actions for this submenu get added in `App:init()`.
![DDF6BB60-7371-472A-8C91-056934279DB7](https://user-images.githubusercontent.com/44080390/163320754-b798975d-986c-4bef-bfc7-9f07a2509d78.jpeg)
The state manager in `statemanager.cpp` has the method `add_state(state)` here:
![CEF12450-9FEC-49D8-9B5C-7205A7C64680](https://user-images.githubusercontent.com/44080390/163320945-c291848d-1747-46f0-a96a-5e5ce41763ea.jpeg)
So it looks like these actions need to be associated with an image from `stockid`, however, `stockid` is deprecated. `StockId's` are initialized in `iconcontroller.cpp`. I am thinking from my research that `Gtk::IconTheme` could replace `stockid`. Here is a snapshot of me trying to test `gtk::IconTheme` but it does not seem to work as I expect:
![2CFE89B3-6B23-4019-92FA-F783B1790B3C](https://user-images.githubusercontent.com/44080390/163321473-a960daa7-3cad-4201-9912-588bf70a2146.jpeg)
The gtkmm documentation says to use `icon names` in place of deprecated `stockid`. Anyways, I am kind of stuck at the moment, any guidance would be appreciagted. Thank you! : )

rodolforg commented:
Please take  a look at #1508

nickleus27 commented:
@rodolforg Since you have already made changes to use `IconTheme` it seems like it would make sense for me not to redundantly make the same changes. Potentially I could either hardcode the images into the `xml menu string` or `xml menu file` like I have done below:
![449C3822-625A-4BCD-AA62-E094405A4EA0_4_5005_c](https://user-images.githubusercontent.com/44080390/163868882-2a9c85ce-c801-4809-94e3-e8c68d1648ba.jpeg)
Or is there a way to pull your draft #1508 so that I can be using the changes you have made for `IconTheme` ?
Currently, I could just keep working on just getting the `menubar` to work by hardcoding .png into `xml` and adding `Gio::SimpleActions`, and not worry about updating `Gtk::Actions` that hold `Gtk::StockIds`.

ice0 commented:
> @rodolforg Since you have already made changes to use `IconTheme` it seems like it would make sense for me not to redundantly make the same changes. Potentially I could either hardcode the images into the `xml menu string` or `xml menu file` like I have done below: 

I think it's ok to hardcode paths on first iteration.


> The gtkmm documentation says to use `icon names` in place of deprecated `stockid`. Anyways, I am kind of stuck at the moment, any guidance would be appreciated. Thank you! : )

I don't know why it doesn't work and I didn't find a good example for an IconTheme.

But, after some testing, this code works for me:

First you need to create `index.theme` file inside `output/Debug/share/synfig/icons/classic/` folder:
```
[Icon Theme]

Name=Classic
Comment=Synfig classic icon theme
Directories=128x128

[128x128]
Size=128
MinSize=12
MaxSize=128
Context=Actions
Type=Scalable
```

Then you need to create a `128x128` folder inside the `classic` folder and copy all the `.png` files into it.

After that change `add_resource_path` to `append_search_path`.

```
auto icon_theme = Gtk::IconTheme::create();
icon_theme->append_search_path(ResourceHelper::get_icon_path());

auto icon = icon_theme->lookup_icon("action_doc_saveas_icon", 128);
if (icon) {
	std::cout << "Found " << icon.get_filename() << "\n";
} else {
	std::cout << "Not found\n";
}
```

Now it should print path to the icon:
```
Found /home/.../synfig/cmake-build-debug/output/Debug/share/synfig/icons/classic/action_doc_saveas_icon.png
```

ice0 commented:
> @ice0 When you have a chance could you check to see if the toggle actions in the `Show/Hide Handles` submenu are working correctly.

I will take a look.

nickleus27 commented:
@ice0 
>But, after some testing, this code works for me:
First you need to create index.theme file inside output/Debug/share/synfig/icons/classic/ folder:

Thank you. This is a very helpful for me to understand how `Icon Theme` works. 
And @rodolforg thank you for sharing your draft of `Icon Theme` .

After some thought I think I am going to continue with these as my steps:

1. hardcoding the `icons` and adding `SimpleActions`
2. once menubar is complete and working move onto removing `Gtk::Actions` that are associated with `StockId`

Let me know if that makes sense.

nickleus27 commented:
what is the `hidden_ui_info_menu` ?

nickleus27 commented:
>But, after some testing, this code works for me:
First you need to create index.theme file inside output/Debug/share/synfig/icons/classic/ folder:

Unfortunately I this did not work for me. I am still getting `not found`.  Not sure why it is not working...

nickleus27 commented:
Hey guys.  I am working on switching the `top_toolbar` in `canvasview.cpp` to the `gio::simpleactions` and `icontheme icons`.  If I use a `.glade` toolbar string in the `Gtk::Builder` object it only allows me to use the corresponding toolbar widget for one container. So when you open new `canvasviews` I cannot call `App::builder()->get_widget("toolbar", toolbar)` to construct the new toolbar in the canvasview.  The method I currently am using I have pushed and I construct each button here: 
![3B1E862F-2708-46CD-A57F-BE7BF3C409D5](https://user-images.githubusercontent.com/44080390/164995967-79fb6d08-6ba1-4e66-b10f-20c3d2885288.jpeg)
The problem is the `Gio::SimpleAction` is not associated with the `button` so when `redo/undo simpleactions` are `disabled` that doesn't correspond to the button. Any ideas how I could do this?

nickleus27 commented:
@ice0 @rodolforg 
Is there anyway I can get help on the different tasks associated with replacing `Gtk::ui_manager` with `Gtk::Builder` because there is so many part to replace: `menus`, `toolbars`, `actions`, `icons`, etc.

ice0 commented:
> @ice0 @rodolforg Is there anyway I can get help on the different tasks associated with replacing `Gtk::ui_manager` with `Gtk::Builder` because there is so many part to replace: `menus`, `toolbars`, `actions`, `icons`, etc. Is trying to do all this in one pull request too much?

Yes, it is better to make only the main menu in this PR.

nickleus27 commented:
@ice0 I have done some work on toolbars and docks. I have updated the `toolbox` , I have updated the `dock_history toolbar` and the as I mentioned half of the `top toolbar`. I did this so I could delete the `gtk::actions` I was replacing with `gio::simpleaction`. My question is can I just leave the changes I made on the toolbars and continue with the `menu`? Or do I need to undo the changes I made to toolbars?

ice0 commented:
> > Yes, it is better to make only the main menu in this PR.
> 
> @ice0 I have done some work on toolbars and docks. I have updated the `toolbox` , I have updated the `dock_history toolbar` and the as I mentioned half of the `top toolbar`. I did this so I could delete the `gtk::actions` I was replacing with `gio::simpleaction`. My question is can I just leave the changes I made on the toolbars and continue with the `menu`? Or do I need to undo the changes I made to toolbars?

Sure, you can leave them. This will need to be done sometime anyway, so it will help a lot.

nickleus27 commented:
@ice0 I cannot find any good documentation on how to add accelerators to a `gtk::menuitem` dynamically.  It says to use set_attribute_value but I cannot get it to work.  Here is an example of how I have used it below, but it does not display the accelerators.

`
menu_item->set_attribute_value(
"accel",Glib::Variant<std::string>::create("&lt;Control&gt;minus"));
`

nickleus27 commented:
@ice0 @rodolforg any ideas on the `toggle_show_menubar` in `mainwindow.cpp`? The toggle button is in the view menu.  With the menubar now being a `gio::menu` and not a `widget` I don't think the menu can `hide()` and `show()` how it did previously?

nickleus27 commented:
@iceo0 @rodolforg I have the menubar done.  There is definitely still some work to be done though.  I think I am missing some things, but I will need your review to help find what is missing.  

1. For example:

![806BB4D0-F91A-4D3D-A5A4-781565B85BDB](https://user-images.githubusercontent.com/44080390/166057203-91ba9f23-523c-43a8-9db4-f6eda0581a50.jpeg)
I havn't found in the codebase where this bottom `section` is added to the `view menu`. The section includes `seek to previous` menu item, etc.  I found this image at https://wiki.synfig.org/Category:View_Menu

2. I still need to find a way to add some `accelerators` dynamically
3. I might be missing an `icon` or two.
4. I am not sure why each time I have been pushing commits the CI tests have been failing. Any ideas?
5. I tried deleting unused code as I replaced deprecated code with new code. However, this potentially was breaking code in other places, so I have had to undo some of my deletions. Please check when reviewing if I should not have deleted something, or if there is code that can be deleted because it is not being used.
6. Should I keep this as a `draft` still or turn this into a `pull request` now?

Thank you for all you help on this!!

ice0 commented on synfig-studio/src/gui/mainwindow.cpp:
```diff
@@ -262,6 +274,18 @@ MainWindow::toggle_show_toolbar()
 }
 
+void
+MainWindow::appmenu_toggle_show_toolbar()
```
You can simplify this by creating helper method:
```
void change_action_state(const std::string& name, bool active) {
	if (auto action = App::instance()->lookup_action(name)) {
		action->change_state(active);
	}
}
```

After that you need only one line to switch it to correct state:
```
void
MainWindow::toggle_show_toolbar()
{
	App::enable_mainwin_toolbar = !App::enable_mainwin_toolbar;
	change_action_state("show-toolbar", App::enable_mainwin_toolbar);
	...
}
```

`App::enable_mainwin_toolbar` already contains the correct state at this point, so you only need to pass it to the action.

Same with MainWindow::toggle_show_menubar:
```
void
MainWindow::toggle_show_menubar()
{
	Gtk::Widget* menubar = App::ui_manager()->get_widget("/menubar-main");
	App::enable_mainwin_menubar = !App::enable_mainwin_menubar;
	change_action_state("show-menubar", App::enable_mainwin_menubar);
	...
}
```

nickleus27 replied:
@ice0 thank you good idea. What do you think about `toggle_show_menu_bar`?  There is nothing to toggle anymore on Macos since the menubar is now not a widget and is in the native `macos` menubar.

ice0 replied:
@nickleus27 
I need to check Linux/Windows behavior. I will answer later.

ice0 commented:
@nickleus27 

Thank you for this great work!

I plan to start merging it step by step after the PR with themes is merged.
This will allow icon names to be used instead of filenames.

Before:
```
<attribute name='icon'>"+App::icon_theme()->lookup_icon(get_icon_name("new"), 128).get_filename()+"</attribute>
```

After:
```
<attribute name='icon'>new</attribute>
```

> any ideas on the toggle_show_menubar in mainwindow.cpp? The toggle button is in the view menu. With the menubar now being a gio::menu and not a widget I don't think the menu can hide() and show() how it did previously?

This menu action can be disabled for MacOS using define:
```
#ifndef __APPLE__
...
#endif
```

There is currently no main menu on Linux, but I plan to solve this issue in the merge step.

nickleus27 commented:
Thanks @ice0 ! That sounds awesome. Also, thank you for explaining how to remove show menubar from apple platform. I just pushed that commit. No need to have `show menubar` in the menu if it doesn't work on `apple`.

nickleus27 commented:
@ice0 I tried building this PR on an linux (ubuntu) docker container and I see what you mean that there is no main menubar on. That is really weird, and I am not sure why that is. I am glad you have an idea how to fix it.  I tried
```
	auto object = App::builder()->get_object("studio_menubar");
	auto gmenu = Glib::RefPtr<Gio::Menu>::cast_dynamic(object);
	auto visible_menubar = Gtk::make_managed<Gtk::MenuBar>(gmenu);
	//auto visible_menubar = App::ui_manager()->get_widget("/menubar-main");
```
but that didn't seem to add the menubar.

ice0 commented:
@nickleus27 

I merged PR with IconTheme support.

You can now use icon names instead of file paths in UI descriptions.

Before:
```
<attribute name='icon'>"+App::icon_theme()->lookup_icon(get_icon_name("quit"), 128).get_filename()+"</attribute>"
```

After:
```
<attribute name='icon'>application-exit</attribute>"
```

This will allow to extract the UI description into an external file and merge it via separate PR (because it won't change the current behavior).

I also found a non-intrusive way to switch from Gtk::Action to Gio::Action. (Currently testing it).

nickleus27 commented:
hey all, when it comes time, I might need some help on how to squash these commits.  Since I have merged with other commits.  Newbie mistake I guess.  Maybe with rebase --interactive?

ice0 commented:
> @ice0 @rodolforg hey all, when it comes time, I might need some help on how to squash these commits. Since I have merged with other commits. Newbie mistake I guess. Maybe with `rebase --interactive`?

Yes you can squash commits using `git rebase --interactive`.
Also there is an awesome GUI tool named SmartGit, it's free for non-commercial use, so I you can try it if you want. 

After you will fix the merge conflicts, this should remove the already merged changes (I mean glade files).

nickleus27 commented:
@ice0 @rodolforg sorry guys I let this commit history get so sloppy.  I downloaded smart git, but I am not sure how to approach cleaning up commit history, because it is not just my commits in a straight line on the graph. I do not want to mess things up.  The history graph has multiple branches, and I am not sure what to do.  Any articles you can direct me to show me how to do this when it is not the last n commits that you are trying to squash.  Thank you.

ice0 commented:
> Also, I am not sure how important it is to clean up history, vs preserving history.

The commit history should help you understand why a particular change was made.
In my opinion, each commit should solve its own problem. If you clean up the code, make a separate commit for this task.
This helps you understand if it's safe to merge this commit without checking it, or if it might break something.
For example this PR: #2638. It can be merged as is because it doesn't change the code.

In the case of this PR, I suppose to leave things as they are. I plan to merge changes from it step by step as separate PRs until all changes are merged.

rodolforg commented:
hi @nickleus27 :) do you still have interest on work on it?
I completely forgot about this amazing work you were doing here. In April and May I was very busy at work, so I couldn't follow this PR.

I have some suggestions here - even some important changes.

nickleus27 commented:
@rodolforg sure whats up?

rodolforg commented:
Sorry my delay. again :(
Before my previous message here, I was working on a draft.
Could you take a look in the last 4 commits of https://github.com/rodolforg/synfig/commits/go-gactions- ?

I created a new class (badly) called ActionManager that is a database for binding action names to labels, (default) shortcuts, etc.
This allows us to save and load custom shortcut configs.

What do you think?

nickleus27 commented:
@rodolforg Right on, this looks cool! Thank you.
